java.lang.RuntimeException: com.intellij.serviceContainer.AlreadyDisposedException: Already disposed: Module: 'prometheus' (disposed) at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:120)

# System

Plugin Version: 14.0.0
Application: IntelliJ IDEA Community Edition (2022.3.1)
Operating System: Mac OS X (13.1)


# Event


## Exception


### Stacktrace


<details>

<summary>
java.lang.RuntimeException: com.intellij.serviceContainer.AlreadyDisposedException: Already disposed: Module: 'prometheus' (disposed)
</summary>

```
java.lang.RuntimeException: com.intellij.serviceContainer.AlreadyDisposedException: Already disposed: Module: 'prometheus' (disposed)
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:120)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:421)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:441)
	at org.elixir_lang.mix.Watcher.syncLibraries(Watcher.kt:70)
	at org.elixir_lang.mix.Watcher.syncLibraries(Watcher.kt:65)
	at org.elixir_lang.mix.Watcher$contentsChanged$1$1.run(Watcher.kt:52)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:423)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$6(CoreProgressManager.java:474)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:188)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:589)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:664)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:620)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:588)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:175)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.intellij.serviceContainer.AlreadyDisposedException: Already disposed: Module: 'prometheus' (disposed)
	at com.intellij.serviceContainer.ComponentManagerImpl.checkState(ComponentManagerImpl.kt:201)
	at com.intellij.serviceContainer.ComponentManagerImpl.getComponent(ComponentManagerImpl.kt:593)
	at com.intellij.openapi.roots.ModuleRootManager.getInstance(ModuleRootManager.java:37)
	at org.elixir_lang.mix.Watcher.syncLibraries$lambda$9$lambda$8(Watcher.kt:74)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:955)
	at org.elixir_lang.mix.Watcher.syncLibraries$lambda$9(Watcher.kt:71)
```

</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -556,6 +556,8 @@
 * [#3042](https://github.com/KronicDeth/intellij-elixir/pull/3042) - [@KronicDeth](https://github.com/KronicDeth)
   * Don't look for variable use scope above EEx tags.
     If a variable is declared in an EEx tag then it is actually missing its declaration and cannot be resolved, so ignore its usage.
+* [#3043](https://github.com/KronicDeth/intellij-elixir/pull/3043) - [@KronicDeth](https://github.com/KronicDeth)
+  * Check if module is disposed when write action is run for `syncLibraries`.
 
 ## v14.0.0
 

--- a/resources/META-INF/changelog.html
+++ b/resources/META-INF/changelog.html
@@ -15,6 +15,7 @@
         Don't look for variable use scope above EEx tags.<br/>
         If a variable is declared in an EEx tag then it is actually missing its declaration and cannot be resolved, so ignore its usage.
       </li>
+      <li>Check if module is disposed when write action is run for <code>syncLibraries</code>.</li>
     </ul>
   </li>
 </ul>

--- a/src/org/elixir_lang/mix/Watcher.kt
+++ b/src/org/elixir_lang/mix/Watcher.kt
@@ -71,61 +71,65 @@ class Watcher(private val project: Project) : BulkFileListener {
                 ApplicationManager.getApplication().runWriteAction {
                     val libraryTable = LibraryTablesRegistrar.getInstance().getLibraryTable(project)
                     val moduleManager = ModuleManager.getInstance(project)
-                    val moduleRootManager = ModuleRootManager.getInstance(module)
 
-                    for (dep in deps) {
-                        if (progressIndicator.isCanceled) {
-                            break
-                        }
+                    if (!module.isDisposed) {
+                        val moduleRootManager = ModuleRootManager.getInstance(module)
 
-                        val depName = dep.application
-
-                        when (dep.type) {
-                            Dep.Type.MODULE -> {
-                                progressIndicator.text2 =
-                                    "Adding $depName Module as dependency of ${module.name} Module"
-
-                                val depModule = moduleManager.findModuleByName(depName)
-
-                                if (depModule != null) {
-                                    if (!moduleRootManager.isDependsOn(depModule)) {
-                                        ModuleRootModificationUtil.addDependency(module, depModule)
-                                    }
-                                } else {
-                                    moduleRootManager.modifiableModel.run {
-                                        addInvalidModuleEntry(depName)
-                                        commit()
-                                    }
-                                }
+                        for (dep in deps) {
+                            if (progressIndicator.isCanceled) {
+                                break
                             }
-                            Dep.Type.LIBRARY -> {
-                                progressIndicator.text2 =
-                                    "Adding $depName Library as dependency of ${module.name} Module"
 
-                                val depLibrary = libraryTable.getLibraryByName(depName)
+                            val depName = dep.application
 
-                                if (depLibrary != null) {
-                                    if (moduleRootManager.orderEntries.none { it is LibraryOrderEntry && it.libraryName == depName }) {
-                                        ModuleRootModificationUtil.addDependency(module, depLibrary)
-                                    }
-                                } else {
-                                    val libraryTableModifiableModule = libraryTable.modifiableModel
+                            when (dep.type) {
+                                Dep.Type.MODULE -> {
+                                    progressIndicator.text2 =
+                                        "Adding $depName Module as dependency of ${module.name} Module"
 
-                                    val invalidLibrary = libraryTableModifiableModule.createLibrary(depName, Kind)
+                                    val depModule = moduleManager.findModuleByName(depName)
 
-                                    libraryTableModifiableModule.commit()
-
-                                    moduleRootManager.modifiableModel.run {
-                                        addLibraryEntry(invalidLibrary)
-                                        commit()
+                                    if (depModule != null) {
+                                        if (!moduleRootManager.isDependsOn(depModule)) {
+                                            ModuleRootModificationUtil.addDependency(module, depModule)
+                                        }
+                                    } else {
+                                        moduleRootManager.modifiableModel.run {
+                                            addInvalidModuleEntry(depName)
+                                            commit()
+                                        }
                                     }
                                 }
 
+                                Dep.Type.LIBRARY -> {
+                                    progressIndicator.text2 =
+                                        "Adding $depName Library as dependency of ${module.name} Module"
+
+                                    val depLibrary = libraryTable.getLibraryByName(depName)
+
+                                    if (depLibrary != null) {
+                                        if (moduleRootManager.orderEntries.none { it is LibraryOrderEntry && it.libraryName == depName }) {
+                                            ModuleRootModificationUtil.addDependency(module, depLibrary)
+                                        }
+                                    } else {
+                                        val libraryTableModifiableModule = libraryTable.modifiableModel
+
+                                        val invalidLibrary = libraryTableModifiableModule.createLibrary(depName, Kind)
+
+                                        libraryTableModifiableModule.commit()
+
+                                        moduleRootManager.modifiableModel.run {
+                                            addLibraryEntry(invalidLibrary)
+                                            commit()
+                                        }
+                                    }
+
+                                }
                             }
                         }
+
+                        syncExternalPathLibraries(deps, progressIndicator)
                     }
-
-                    syncExternalPathLibraries(deps, progressIndicator)
                 }
             }
         }

